File: public/methods/40_PORTFOLIO_METHODS/QM008_DRAWDOWN_UNDERWATER_DURATION_RECOVERY/labs/python/qm008_hands_on.py
from pathlib import Path
import csv

TOL = 1e-12
root = Path(__file__).resolve().parents[1]
with (root / 'data/qm008_wealth_path.csv').open(newline='', encoding='utf-8') as f:
    rows = [(int(r['t']), float(r['wealth'])) for r in csv.DictReader(f)]

high = -float('inf')
last_high = rows[0][0]
states = []
for t, wealth in rows:
    if wealth >= high - TOL:
        if wealth > high + TOL:
            high = wealth
        last_high = t
        dd = 0.0
        age = 0
    else:
        dd = 1.0 - wealth / high
        age = t - last_high
    states.append((t, wealth, high, dd, -dd, age))

print('t wealth hwm drawdown underwater age')
for row in states:
    print(f'{row[0]:2d} {row[1]:6.1f} {row[2]:6.1f} {row[3]:8.4f} {row[4]:10.4f} {row[5]:3d}')
print(f'maximum drawdown={max(x[3] for x in states):.6f}')
print(f'terminal censored={str(states[-1][3] > TOL).lower()}')

File: public/methods/40_PORTFOLIO_METHODS/QM008_DRAWDOWN_UNDERWATER_DURATION_RECOVERY/labs/r/qm008_hands_on.R
args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
script_path <- if (length(file_arg)) sub("^--file=", "", file_arg[1]) else normalizePath("qm008_hands_on.R")
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)
df <- read.csv(file.path(root, "data", "qm008_wealth_path.csv"))

tol <- 1e-12
high <- -Inf
last_high <- df$t[1]
out <- data.frame()
for (k in seq_len(nrow(df))) {
  t <- df$t[k]
  w <- df$wealth[k]
  if (w >= high - tol) {
    if (w > high + tol) high <- w
    last_high <- t
    dd <- 0
    age <- 0
  } else {
    dd <- 1 - w/high
    age <- t - last_high
  }
  out <- rbind(out, data.frame(t=t, wealth=w, hwm=high, drawdown=dd, underwater=-dd, age=age))
}
print(out, row.names=FALSE)
cat(sprintf("maximum drawdown=%.6f\n", max(out$drawdown)))
cat(sprintf("terminal censored=%s\n", tolower(as.character(tail(out$drawdown,1) > tol))))
